// stylelint/dist/rules/selector-pseudo-class-case/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (expectation) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["lower", "upper"]
    });
    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {
      var selector = rule.selector;
      var startIndexPseudo = selector.indexOf(":");

      if (startIndexPseudo === -1) {
        return;
      }

      (0, _postcssSelectorParser2.default)(function (selectorTree) {
        selectorTree.eachPseudo(function (pseudoNode) {
          var pseudo = pseudoNode.value;

          if (pseudo.indexOf("::") !== -1 || pseudoElements.has(pseudo.toLowerCase())) {
            return;
          }

          var expectedPseudo = expectation === "lower" ? pseudo.toLowerCase() : pseudo.toUpperCase();

          if (pseudo === expectedPseudo) {
            return;
          }

          (0, _utils.report)({
            message: messages.expected(pseudo, expectedPseudo),
            node: rule,
            index: pseudoNode.sourceIndex,
            ruleName: ruleName,
            result: result
          });
        });
      }).process(selector);
    });
  };
};

var _postcssSelectorParser = require("postcss-selector-parser");

var _postcssSelectorParser2 = _interopRequireDefault(_postcssSelectorParser);

var _utils = require("../../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ruleName = exports.ruleName = "selector-pseudo-class-case";

var messages = (0, _utils.ruleMessages)(ruleName, {
  expected: function expected(actual, _expected) {
    return "Expected \"" + actual + "\" to be \"" + _expected + "\"";
  }
});

// level 1 and 2 pseudo elements
exports.messages = messages;
var pseudoElements = new Set([":before", ":after", ":first-line", ":first-letter"]);
